packages/flutter/lib/src/rendering/object.dart

```diff
@@ -2580,6 +2580,16 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
         node = node.parent;
       } while (node._semantics == null);
       node._semantics?.reset();
+      if (node != this && _semantics != null && _needsSemanticsUpdate) {
+        // If [this] node has already been added to [owner._nodesNeedingSemantics]
+        // remove it as it is no longer guaranteed that its semantics
+        // node will continue to be in the tree. If it still is in the tree, the
+        // ancestor [node] added to [owner._nodesNeedingSemantics] at the end of
+        // this block will ensure that the semantics of [this] node actually get
+        // updated.
+        // (See semantics_10_test.dart for an example why this is required).
+        owner._nodesNeedingSemantics.remove(this);
```

Comment:
Some alternatives:
1) We can remember the "removed" nodes in a separate list and check agains that in `flushSemantics`. That would just move the O(N) to `flushSemantics` though.
2) We can change `_nodesNeedingSemantics` to a Set, which would give us O(1) for add and remove. However, that will add an extra conversion from Set to some kind of an ordered collection at the beginning of the `flushSemantics` phase. I imagine that conversion is O(N), but it would only happen once per frame instead of multiple times, so probably more efficient.
3) Remove the assert and instead replace it with logic that drops the node on the floor since that semantic node is no longer needed. That sounds dangerous as that assert might catch other error cases and at the point where the assert fires, we've already done a lot of (now useless) work to get the node into the tree.

Of those options, 2) seems the most reasonable. I refactored it to that. Let me know what you think.